tests/AggregationBlogPost/article.md

```diff
@@ -0,0 +1,713 @@
+but they allow us to know what types an expression will resolve to.
+
+For an example, let's take the `$hour` operator. When you use
+`Expression::hour`, you will receive an instance of an `HourOperator` class.
+This class implements an `OperatorInterface`, telling the builder that this is
+an operator that can be used in aggregation pipeline. It also implements a
+`ResolvesToInt` interface, as we always know that evaluating the expression
+results in an integer value. The required `date` parameter of the operator is a
+date, which in the builder is one of many things. It could be a BSON
+`UTCDateTime` instance, but it could also be the result of any operator that
+returns a date, e.g. `$dateFromString`.
+
+Now that we know about these value holder objects, we still need to make sure
+the server knows what we're talking about. When you call `Collection::aggregate`
+with the pipeline you built, what happens internally to it? Here, a series of
+encoders springs into action. We use a single entry point, the `BuilderEncoder`
+class. This class contains multiple encoders that are able to handle all
+pipeline stages, operators, and accumulators and transform them into their BSON
+representations.
+
+This allows us to keep the logic customizable. For example, the Doctrine MongoDB
+ODM allows users to specify different names for fields in the database. In turn,
+it needs to convert the name of a property in the mapped class to the name of
+the field in the database. Such a feature could easily be built by creating a
+custom encoder for all `fieldPath` expressions, and changing the field path
+accordingly.
+
+When creating a `MongoDB\Client` instance, you can now pass an additional
+`builderEncoder` option in the `$driverOptions` argument. This specifies the
+encoder used to encode aggregation pipelines, but also query objects. All
+`Database` and `Client` instances inherit this value from the client, but you
+can override it through the options when fetching such an instance. This allows
+you to have your custom logic applied whenever pipelines or queries are encoded
+for the server.
+
+With factories, value holders, and encoders, we wanted to ensure that creating
+the builder does not turn into a repetitive chore. As you can imagine, many
+operators will mostly consist of the same logic, resulting in tons of code
+duplication. To make matters worse, every new server version adds some new
+operators or even stages, so we wanted to make sure that we can easily expand
+the builder.
+
+We could try to rely on generative AI to help us with this, but this only goes
+so far. Instead, we leverage code generation to make the task easier. All
+factories, value holders, and encoders are generated from a configuration. When
+a new operator is introduced, we create a config file with all of its details:
+input types, what the operator resolves to, documentation for parameters, and
+even the examples from the documentation are included. We then run the
```

Comment:
```suggestion
even examples from the MongoDB documentation. We then run the
```

This part of the sentence is a list (following "details:") noting what's included, so "are included" seemed redundant.